### PR TITLE
[Improvement](ObjectStorage) Retry when encountering TooManyRequest response

### PR DESCRIPTION
**Pull request by ByteYue**
## Proposed changes

We can do retry for 429 using the embedded utility of corresponding sdk.

### CODE REVIEW

**Comment by ByteYue** [comment]
run buildall

File: be/src/io/fs/s3_file_reader.cpp [hunk 1]
```diff
@@ -187,16 +139,10 @@ void S3FileReader::_collect_profile_before_close() {
         ADD_TIMER(_profile, s3_profile_name);
         RuntimeProfile::Counter* total_get_request_counter =
                 ADD_CHILD_COUNTER(_profile, "TotalGetRequest", TUnit::UNIT, s3_profile_name);
-        RuntimeProfile::Counter* too_many_request_err_counter =
```
Comment (gavinchou) [inline be/src/io/fs/s3_file_reader.cpp:190]: do not remove these profile items

**Review by gavinchou — CHANGES_REQUESTED** [review]
pls retain the profile items

File: be/src/util/s3_util.cpp [hunk 2]
```diff
@@ -73,6 +73,20 @@ bool to_int(std::string_view str, int& res) {
...
 
+class CustomRetryStrategy final : public Aws::Client::DefaultRetryStrategy {
+public:
+    CustomRetryStrategy(int maxRetries) : DefaultRetryStrategy(maxRetries) {}
+
+    bool ShouldRetry(const Aws::Client::AWSError<Aws::Client::CoreErrors>& error,
+                     long attemptedRetries) const override {
+        if (attemptedRetries < m_maxRetries &&
+            error.GetResponseCode() == Aws::Http::HttpResponseCode::TOO_MANY_REQUESTS) {
+            return true;
```
Comment (gavinchou) [inline be/src/util/s3_util.cpp:84]: We should record something here if we are doing retry.
e.g. a retry counter or something useful for us to observe the status of requesting the object storage service.

File: be/src/io/fs/s3_file_reader.cpp [hunk 3]
```diff
@@ -120,65 +112,25 @@ Status S3FileReader::read_at_impl(size_t offset, Slice result, size_t* bytes_rea
...
-        s3_bytes_read_total << bytes_req;
-        s3_bytes_per_read << bytes_req;
-        DorisMetrics::instance()->s3_bytes_read_total->increment(bytes_req);
-        if (retry_count > 0) {
-            LOG(INFO) << fmt::format("read s3 file {} succeed after {} times with {} ms sleeping",
-                                     _path.native(), retry_count, total_sleep_time);
-        }
-        return Status::OK();
+    // clang-format off
+    auto resp = client->get_object( { .bucket = _bucket, .key = _key, },
```
Comment (gavinchou) [inline be/src/io/fs/s3_file_reader.cpp:116]: Considering create a new client with `CustomRetryStrategy` which can carry enough context
such as number retires and duration of all retires.

**Review by gavinchou — CHANGES_REQUESTED** [review]

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run beut

**Comment by ByteYue** [comment]
run feut

**Comment by ByteYue** [comment]
run cloudut

**Comment by ByteYue** [comment]
run buildall

**Review by platoneko — DISMISSED** [review]

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

File: common/cpp/obj_retry_strategy.cpp [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+// specific language governing permissions and limitations
+// under the License.
+
+#include "obj_retry_strategy.h"
+
+#include <bvar/reducer.h>
+
+namespace doris {
+
+bvar::Adder<int64_t> too_many_request_retry_times("too_many_request_retry_times");
```
Comment (gavinchou) [inline common/cpp/obj_retry_strategy.cpp:24]: Naming s3_too_many_request_retry_cnt

**Review by gavinchou — CHANGES_REQUESTED** [review]

**Comment by ByteYue** [comment]
run buildall

**Review by gavinchou — APPROVED** [review]

**Review by Yukang-Lian — APPROVED** [review]

**Review by platoneko — APPROVED** [review]

**Pull request merged**